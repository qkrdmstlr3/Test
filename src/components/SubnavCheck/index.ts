import {
  ShellHTML,
  createComponent,
  EventType,
  RenderType,
  useGlobalState,
  setGlobalState,
} from '@Lib/shell-html';
import { CheckListType, CheckPostType } from '@Types/types';
import styleSheet from './style.scss';
import { ipcRenderer } from 'electron';
import short from 'short-uuid';
import { CheckPostStatusType } from '../../types/enum';
import { getDday } from '../../utils/calcDate';

class SubnavCheck extends ShellHTML {
  constructor() {
    super({
      selectedItem: useGlobalState('checkpostControl').currentCheckListId,
      toModifyItem: false,
    });
  }

  connectedCallback(): void {
    this.enrollObserving('checklist');
    this.enrollObserving('checkpostControl');

    this.getList();
  }

  disconnectedCallback(): void {
    this.releaseObserving('checklist');
    this.releaseObserving('checkpostControl');
    ipcRenderer?.removeAllListeners('checklist:read:all');
  }

  getList(): void {
    const list = useGlobalState('checklist');
    if (!list.length) {
      ipcRenderer?.send('checklist:read:all');
      ipcRenderer?.once('checklist:read:all', (event, list) => {
        setGlobalState('checklist', list);
      });
    }
  }

  clickAccordionHeaderHandler(event: Event): void {
    if (!(event.target instanceof HTMLElement)) return;

    const accordionItemId = event.target.closest('.accordion__header')?.id;
    if (this.state.selectedItem !== accordionItemId) {
      this.setState({ selectedItem: accordionItemId, toModifyItem: false });
    } else if (this.state.selectedItem.length && !this.state.toModifyItem) {
      this.setState({ selectedItem: undefined, toModifyItem: false });
    }
  }

  clickAddButtonHandler(event: Event): void {
    if (!(event.target instanceof HTMLElement)) return;

    const list = useGlobalState('checklist');
    const newItem = { id: short.generate(), name: '새 목록' };

    setGlobalState('checklist', [...list, { ...newItem, posts: [] }]);
    ipcRenderer?.send('checklist:add', newItem);
  }

  modifyButtonHandler(event: Event): void {
    if (!(event.target instanceof HTMLElement)) return;

    const accordionItemId = event.target.closest('.accordion__header')?.id;
    if (!this.state.toModifyItem) {
      this.setState({
        selectedItem: accordionItemId,
        toModifyItem: true,
      });
    }
  }

  submitToChangeNameHandler(event: Event): void {
    event.preventDefault();
    if (!(event.target instanceof HTMLFormElement)) return;

    // queryselector사용을 바꾸면 좋을 것 같은데...
    const newName = event.target.querySelector('input')?.value;
    if (!newName || !newName.length) return; //TODO: alert

    const list = useGlobalState('checklist');
    list.forEach((item: CheckListType) => {
      if (item.id === this.state.selectedItem) {
        item.name = newName;
      }
    });

    ipcRenderer?.send('checklist:modify:name', {
      id: this.state.selectedItem,
      newName: newName,
    });
    setGlobalState('checklist', list);
    this.setState({ ...this.state, toModifyItem: false });
  }

  deleteItemHandler(event: Event): void {
    event.preventDefault();
    if (!(event.target instanceof HTMLElement)) return;

    const deleteMessage =
      '하위 글들이 모두 사라집니다.\n 정말 삭제하시겠습니까? ';
    const isDelete = confirm(deleteMessage);
    if (!isDelete) return;

    const list = useGlobalState('checklist');
    setGlobalState(
      'checklist',
      list.filter((item: CheckListType) => item.id !== this.state.selectedItem)
    );
    ipcRenderer?.send('checklist:delete', {
      id: this.state.selectedItem,
    });
  }

  createNewCheckPost(): CheckPostType {
    return {
      id: short.generate(),
      title: '새 게시글',
      status: CheckPostStatusType.todo,
      endDate: '',
      startDate: '',
      content: `
        <h1 contenteditable="true">제목</h1>
        <div contenteditable="true">내용</div>
      `,
    };
  }

  createCheckPostHandler(event: Event): void {
    event.preventDefault();
    // 게시글 id를 전역으로 관리할까말까..? > 하자!
    if (!this.state.selectedItem) {
      alert('게시글이 들어갈 리스트를 선택해주십시오');
      return;
    }

    const newCheckPost = this.createNewCheckPost();
    const checkposts = useGlobalState('checkposts');
    const checkpostControl = useGlobalState('checkpostControl');
    const list = useGlobalState('checklist');
    list.forEach((item: CheckListType) => {
      if (item.id === this.state.selectedItem) {
        item.posts.push({
          id: newCheckPost.id,
          title: newCheckPost.title,
          dday: getDday(newCheckPost.endDate),
        });
      }
    });
    ipcRenderer.send('checkpost:add', {
      ...newCheckPost,
      listId: this.state.selectedItem,
    });
    setGlobalState('checkposts', [...checkposts, newCheckPost]);
    setGlobalState('checkpostControl', {
      ...checkpostControl,
      currentCheckPostId: newCheckPost.id,
    });
    setGlobalState('checklist', list);
  }

  clickCheckPostHandler(event: Event): void {
    event.preventDefault();
    if (!(event.target instanceof HTMLElement)) return;

    const checkPostId = event.target.closest('.accordion__item')?.id;
    const checkpostControl = useGlobalState('checkpostControl');
    if (checkPostId === checkpostControl.currentCheckPostId) return;
    setGlobalState('checkpostControl', {
      ...checkpostControl,
      currentCheckPostId: checkPostId,
    });
  }

  render(): RenderType {
    const list = useGlobalState('checklist');
    const { currentCheckPostId } = useGlobalState('checkpostControl');
    const listHTML = list.reduce(
      (acc: string, { id, name, posts }: CheckListType) =>
        (acc += `
      <li class="accordion">
        <header class="accordion__header ${
          this.state.selectedItem === id ? 'choosed__list' : ''
        }" id="${id}" data-testid="${id}">
          ${
            this.state.toModifyItem && this.state.selectedItem === id
              ? `
              <form class="accordion__form" data-testid="test_form">
                <input 
                  placeholder="목록 이름" 
                  value="${name}" 
                  class="accordion__input" 
                  name="listname" 
                  maxlength="13"
                  data-testid="test_input"
                />
              </form>`
              : `<h3 class="accordion__name">${name}</h3>`
          }
          <div class="accordion__header__buttons">
          ${
            this.state.toModifyItem && this.state.selectedItem === id
              ? ''
              : `<button class="accordion__button__modify" data-testid="modify_button">ℳ</button>
            <button class="accordion__button__delete" data-testid="delete_button">✖️</button>`
          }
          </div>
        </header>
        ${
          this.state.selectedItem === id && posts.length
            ? `<ul class="accordion__list">
            ${posts.reduce(
              (acc, item) =>
                (acc += `
              <div class="accordion__item ${
                currentCheckPostId === item.id ? 'choosed__item' : ''
              }" id="${item.id}">
                <span>${item.title}</span>
                <span>${item.dday}</span>
              </div>
              `),
              ''
            )}
          </ul>`
            : ''
        }
      </li>
      `),
      ''
    );

    return {
      css: styleSheet,
      eventFuncs: [
        {
          className: 'accordion__button__delete',
          func: this.deleteItemHandler,
          type: EventType.click,
        },
        {
          className: 'accordion__header',
          func: this.clickAccordionHeaderHandler,
          type: EventType.click,
        },
        {
          className: 'accordion__button__modify',
          func: this.modifyButtonHandler,
          type: EventType.click,
        },
        {
          className: 'accordion__form',
          func: this.submitToChangeNameHandler,
          type: EventType.submit,
        },
        {
          className: 'accordion__plus',
          func: this.clickAddButtonHandler,
          type: EventType.click,
        },
        {
          className: 'nav__top__button',
          func: this.createCheckPostHandler,
          type: EventType.click,
        },
        {
          className: 'accordion__item',
          func: this.clickCheckPostHandler,
          type: EventType.click,
        },
      ],
      html: `
      <nav class="nav">
        <div class="nav__top">
          <h1 class="nav__top__title">할 일들</h1>
          <button class="nav__top__button">+</button>
        </div>
        <ul class="nav__bottom">
          ${listHTML}
          <li class="accordion__plus" data-testid="add_button"> + </li>
        </ul>
      </nav>
      `,
    };
  }
}

createComponent('subnav-check', SubnavCheck);

export default SubnavCheck;
