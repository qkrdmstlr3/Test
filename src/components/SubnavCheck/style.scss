@import '/src/common/style/mixin.scss';
@import '/src/common/style/variables.scss';

.nav {
  @include subnav;
  padding: 0 1rem;
  box-sizing: border-box;
  user-select: none;

  &__top {
    display: flex;
    justify-content: space-between;
    align-items: center;

    &__title {
      font-size: 1.8rem;
    }

    &__button {
      @include center;
      width: 2rem;
      height: 2rem;
      padding: 0;
      border-radius: 1rem;
      outline: none;
      font-size: 2rem;
      border: 2px solid white;

      color: $deep-pink-color;
      background-color: white;
      border-left-color: $pink-color;
      border-top-color: $pink-color;
      border-right-color: $deep-pink-color;
      border-bottom-color: $deep-pink-color;
      cursor: pointer;

      &:active {
        border-left-color: $deep-pink-color;
        border-top-color: $deep-pink-color;
        border-right-color: $pink-color;
        border-bottom-color: $pink-color;
      }
    }
  }

  &__bottom {
    margin: 0;
    padding: 0;
    list-style: none;
  }
}

.accordion {
  &__plus,
  &__header,
  &__item {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin: 0.2rem 0;
    padding: 0.5rem;

    border: 1px solid $pink-color;
    border-radius: 5px;
    color: $black-color;
    cursor: pointer;

    &:hover {
      background-color: $pink-color;
    }
  }

  &__header {
    height: 1.5rem;
    color: $deep-pink-color;
    border: 1px solid $deep-pink-color;

    &:hover {
      color: white;
      background-color: $deep-pink-color;
      .accordion__header__buttons {
        display: block;
      }
    }

    &__buttons {
      display: none;
      button {
        background-color: transparent;
        border: none;
        color: white;
        cursor: pointer;
        outline: none;
      }
    }
  }

  &__plus {
    @include center;
    padding: 0;

    color: white;
    background-color: $pink-color;
    font-weight: bold;
    font-size: 2rem;

    &:hover {
      background-color: $deep-pink-color;
    }
  }

  &__name,
  &__item__name {
    margin: 0;
  }

  &__list {
    margin: 0;
    padding: 0;
    list-style: none;
    overflow: hidden;
  }

  &__input {
    padding: none;
    border: none;

    color: white;
    font-weight: bold;
    border-bottom: 1px solid white;
    background-color: $deep-pink-color;
    outline: none;

    &::placeholder {
      color: white;
      opacity: 0.8;
    }
  }

  &__item {
    &:last-child {
      margin-bottom: 1.5rem;
    }
  }

  &__status {
    margin-right: 0.2rem;
    font-weight: bold;
    color: $deep-pink-color;
  }
}

.choosed {
  &__list {
    color: white;
    background-color: $deep-pink-color;
  }

  &__item {
    background-color: $pink-color;
  }
}
