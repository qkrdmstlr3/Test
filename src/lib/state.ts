import { state } from '@Lib/shell-html';

state({
  key: 'page',
  initial: 'home',
});

state({
  key: 'checklist',
  initial: [],
});

state({
  key: 'checkpostControl',
  initial: {
    currentCheckListId: undefined,
    currentCheckPostId: undefined,
  },
});

state({
  key: 'checkposts',
  initial: [],
});
