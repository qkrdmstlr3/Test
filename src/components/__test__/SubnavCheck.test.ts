/* dependencies */
import { fireEvent, getByTestId, getByText } from '@testing-library/dom';
import '@testing-library/jest-dom/extend-expect';
import _ from 'lodash';

/* component, state */
import SubnavCheck from '../SubnavCheck';
import '@Lib/state';
import { useGlobalState, setGlobalState } from 'shell-html';

let subnav: SubnavCheck;
let subnavComponent: HTMLElement;

window.confirm = jest.fn(() => true);

const dummyChecklist = [
  { id: 'list1', name: '리스트1', posts: [] },
  { id: 'list2', name: '리스트2', posts: [] },
  { id: 'list3', name: '리스트3', posts: [] },
];

describe('SubnavCheck Component test', () => {
  beforeEach(() => {
    subnav = new SubnavCheck();
    subnav.connectedCallback();
    if (subnav.shadowRoot?.childNodes[0]) {
      subnavComponent = (subnav.shadowRoot as unknown) as HTMLElement;
    }
    setGlobalState('checklist', _.cloneDeep(dummyChecklist));
  });

  afterEach(() => {
    subnav.disconnectedCallback();
  });

  it('list rendering test', () => {
    if (!subnavComponent) return;

    dummyChecklist.forEach((item) => {
      const list = getByTestId(subnavComponent, item.id);
      expect(list.innerHTML).toContain(item.name);
    });
  });

  it('select item test', () => {
    if (!subnavComponent) return;

    fireEvent.click(getByTestId(subnavComponent, dummyChecklist[0].id), {
      button: 0,
    });
    const firstItem = getByTestId(subnavComponent, dummyChecklist[0].id);
    const secondItem = getByTestId(subnavComponent, dummyChecklist[1].id);
    const thirdItem = getByTestId(subnavComponent, dummyChecklist[2].id);

    expect(subnav.state.selectedItem).toBe('list1');
    expect(firstItem.classList.contains('choosed__list')).toBeTruthy();
    expect(secondItem.classList.contains('choosed__list')).toBeFalsy();
    expect(thirdItem.classList.contains('choosed__list')).toBeFalsy();
  });

  it('create new item test', () => {
    if (!subnavComponent) return;

    fireEvent.click(getByTestId(subnavComponent, 'add_button'));
    const lists = useGlobalState('checklist');
    const a = getByText(subnavComponent, '새 목록');

    expect(lists.length).toBe(4);
    expect(a.innerHTML).toContain('새 목록');
  });

  it('change item name test', () => {
    if (!subnavComponent) return;

    const firstItem = getByTestId(subnavComponent, dummyChecklist[0].id);
    const modifyButton = getByTestId(firstItem, 'modify_button');
    fireEvent.click(modifyButton as Element);

    const input = getByTestId(
      subnavComponent.childNodes[0] as HTMLElement,
      'test_input'
    );
    expect(input instanceof HTMLInputElement).toBeTruthy();
    (input as HTMLInputElement).value = '테스트 목록';

    const form = getByTestId(subnavComponent, 'test_form');
    expect(form instanceof HTMLFormElement).toBeTruthy();
    fireEvent.submit(form);

    expect(
      getByTestId(subnavComponent, dummyChecklist[0].id).innerHTML
    ).toContain('테스트 목록');
  });

  it('delete item test', () => {
    if (!subnavComponent) return;

    fireEvent.click(getByTestId(subnavComponent, dummyChecklist[0].id));
    const firstItem = getByTestId(subnavComponent, dummyChecklist[0].id);
    const deleteButton = getByTestId(firstItem, 'delete_button');

    fireEvent.click(deleteButton);
    expect(window.confirm).toBeCalled();

    expect(
      subnavComponent.innerHTML.includes(dummyChecklist[0].name)
    ).toBeFalsy();
  });

  it('snapshot', () => {
    expect(subnavComponent).toMatchSnapshot();
  });
});
