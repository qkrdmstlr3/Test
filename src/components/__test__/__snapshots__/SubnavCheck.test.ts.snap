// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubnavCheck Component test snapshot 1`] = `
ShadowRoot {
  Symbol(SameObject caches): Object {
    "childNodes": NodeList [
      <nav
        class="nav"
      >
        <div
          class="nav__top"
        >
          <h1
            class="nav__top__title"
          >
            할 일들
          </h1>
          <button
            class="nav__top__button"
          >
            +
          </button>
        </div>
        <ul
          class="nav__bottom"
        >
          <li
            class="accordion"
          >
            <header
              class="accordion__header "
              data-testid="list1"
              id="list1"
            >
              <h3
                class="accordion__name"
              >
                리스트1
              </h3>
              <div
                class="accordion__header__buttons"
              >
                <button
                  class="accordion__button__modify"
                  data-testid="modify_button"
                >
                  ℳ
                </button>
                <button
                  class="accordion__button__delete"
                  data-testid="delete_button"
                >
                  ✖️
                </button>
              </div>
            </header>
          </li>
          <li
            class="accordion"
          >
            <header
              class="accordion__header "
              data-testid="list2"
              id="list2"
            >
              <h3
                class="accordion__name"
              >
                리스트2
              </h3>
              <div
                class="accordion__header__buttons"
              >
                <button
                  class="accordion__button__modify"
                  data-testid="modify_button"
                >
                  ℳ
                </button>
                <button
                  class="accordion__button__delete"
                  data-testid="delete_button"
                >
                  ✖️
                </button>
              </div>
            </header>
          </li>
          <li
            class="accordion"
          >
            <header
              class="accordion__header "
              data-testid="list3"
              id="list3"
            >
              <h3
                class="accordion__name"
              >
                리스트3
              </h3>
              <div
                class="accordion__header__buttons"
              >
                <button
                  class="accordion__button__modify"
                  data-testid="modify_button"
                >
                  ℳ
                </button>
                <button
                  class="accordion__button__delete"
                  data-testid="delete_button"
                >
                  ✖️
                </button>
              </div>
            </header>
          </li>
          <li
            class="accordion__plus"
            data-testid="add_button"
          >
             + 
          </li>
        </ul>
      </nav>,
      <style>
        [object Object]
      </style>,
    ],
  },
}
`;
